[Storage] Fix MD5 not work on FIPS enabled machine issue

blueww opened the pull request:
<!-- DO NOT DELETE THIS TEMPLATE -->

## Description

<!-- Please add a brief description of the changes made in this PR -->

Fix the issue  https://github.com/Azure/azure-powershell/issues/4807

The resolution:

- At the begin of the related cmdlets, try call the MD5 caculate function, if the MD5 function report FIPS exception, set DMlib to use native MD5; else still use the orignal MD5 algrithm.

With this change:

- we can get the issue resolved without adding new parameter, so script can be shared across machines without change
- PowerShell by default will still use the orignal MD5 algrithm, so aligned with .net SDK  (the MD5 check cost will be very low)
- It don’t need big cmdlet framework change.



## Checklist

- [x] I have read the [_Submitting Changes_](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#making-changes) section of [`CONTRIBUTING.md`](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md)
- [x] The title of the PR is clear and informative
- [x] The appropriate [change log has been updated](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#updating-the-change-log)
- [x] The PR does not introduce [breaking changes](https://github.com/Azure/azure-powershell/blob/preview/documentation/breaking-changes/breaking-changes-definition.md)
- [x] If applicable, the changes made in the PR have proper test coverage
- [x] For public API changes to cmdlets:
    - [x] the changes have gone through a [cmdlet design review](https://github.com/Azure/azure-powershell-cmdlet-review-pr)
    - [x] the cmdlet markdown files were [generated using the `platyPS` module](https://github.com/Azure/azure-powershell/blob/preview/documentation/development-docs/help-generation.md)

blueww commented:
@praries880 

Currently, we don't have Unit test for E2E of Storage data plan cmdlets. But will cover it in another test framework. 
As I know, @MiYanni might will add Unit test for it in the future.  

So for this change, the test we can do is to check FIPSEnabled is disable on normal machine. But dur to the protect level, we can't access FIPSEnabled from test case. I can change FIPSEnabled to public to add a test. But it looks not proper to change an object from protected to public just for test.

So I think we can wait for @MiYanni to add the storage Unit test, then add test case for it.

MiYanni commented:
@blueww I'm not involved in this situation/issue. What's going on?

blueww commented:
@MiYanni  
Don't worry. It's just I am asked to add Unit test for some Storage data plan cmdlets change. But we don't have E2E Storage Dataplan test in PowerShell Unit test now.
As I remember, Mark send you a mail to add Storage Data plan test based on a script before. 
So this test should base on the test change that change.
I am not pushing you to add the test now. Just explain why we can't add test for this change now.

cormacpayne commented:
Sign job: https://azuresdkci.westus2.cloudapp.azure.com/job/ps-sign/265/

praries880 commented:
@blueww   Thanks for the explanation. 
Since we cant create a test, would you mind testing the signed bits for your changes (found here https://azuresdkci.westus2.cloudapp.azure.com/job/ps-sign/265/artifact/src/Package/) and test the functionality on a machine where you can controll the FIPS setting?

markcowl commented on src/Storage/Commands.Storage/Common/StorageCloudCmdletBase.cs:
```diff
@@ -622,5 +622,32 @@ protected override void StopProcessing()
+                return false;
+            }
+        }
+        private static bool? fIPSEnabled = null;
```
variable name should be fipsEnabled

blueww replied:
fixed, use fipsEnabled

markcowl commented on src/Storage/Commands.Storage/Common/StorageCloudCmdletBase.cs:
```diff
@@ -622,5 +622,32 @@ protected override void StopProcessing()
+                    fIPSEnabled = true;
+                    return true;
+                }
+                fIPSEnabled = false;
```
vertical space

markcowl replied:
```c#
public static bool FipsEnabled {get;} = IsFIPSEnabled();

internal static bool ISFIPSEnabled()
{
                try
                {
                    System.Security.Cryptography.MD5.Create();
                    return false;
                }
              catch (System.Reflection.TargetInvocationException)
                {
                    return true;
                }
}
```

markcowl replied:
I think we can take advantage of the fact that the static initializer will run once to simploify the code - so we cans et the value and then use it later

blueww replied:
Fixed as the code from you. Thanks!

markcowl commented on src/Storage/Commands.Storage/File/Cmdlet/SetAzureStorageFileContent.cs:
```diff
@@ -80,6 +80,12 @@ public override void ExecuteCmdlet()
+            // if FIPS policy is enabled, must use native MD5
+            if (FIPSEnabled)
+            {
+                CloudStorageAccount.UseV1MD5 = false;
```
Is this the only cmdlet that would be impacted?  Wouldn't every blob-related cmdlet need this setting?

blueww replied:
Only upload blob/file cmdlets need use MD5 function to calculate Property ContentMD5. Other cmdlets don't use this function. So only need change cmdlets Set-AzureStorageBlob(File)Content

blueww commented on src/Storage/Commands.Storage/Common/StorageCloudCmdletBase.cs:
```diff
@@ -622,5 +622,32 @@ protected override void StopProcessing()
+        {
+            get
+            {
+                if (fIPSEnabled.HasValue)
```
fixed, use fipsEnabled

blueww commented:
@markcowl 

I have fixed the comments from you, please help to check. Change log is also added.
The on-demand test can be found in: https://azuresdkci.westus2.cloudapp.azure.com/job/powershell-demand/650/
